doc/neps/nep-0029-dtype-as-type.rst

```diff
@@ -0,0 +1,286 @@
+                    # Dtype('int8') or Dtype('S10') or record descr
+                    return create_new_descr(cls, *args, **kwargs)
+            else:
+                # At import, when creating Dtype and subclasses
+                return type.__call__(cls, *args, **kwargs)
+
+    class Dtype():
+        def __new__(cls, *args, **kwargs):
+            # Do not allow creating instances of abstract base classes
+            if not hasattr(cls, 'itemsize'):
+                raise ValueError("cannot create instances of "
+                                 f"abstract class {cls!r}")
+            return super().__new__(cls, *args, **kwargs)
+
+    class GenericDescr(Dtype, metaclass=DtypeMeta):
+        pass
+
+    class IntDescr(GenericDescr):
+        def __repr__(self):
+            # subclass of IntDescr
+            return f"dtype('{_kind_to_stem[self.kind]}{self.itemsize:d}')"
+
+        def get_format_function(self, data, **options):
+            # replaces switch on dtype found in _get_format_function
+            # (in arrayprint), **options details missing
+            from  np.core.arrayprint import IntegerFormat
+            return IntegerFormat(data)
+
+    class UInt8Descr(IntDescr):
+        kind = 'u'
+        itemsize = 8
+        type = np.uint8
+        # The C-based methods for sort, fill, cast, clip, ... not exposed to
+        # Python
+        #ArrFuncs = int8_arrayfuncs
+        
+
+    dtype_int_dict = {1: UInt8Descr()}
+    dtype_scalar_dict = {np.uint8: UInt8Descr()}
+    _kind_to_stem = {
+        'u': 'uint',
+        'i': 'int',
+        'c': 'complex',
+        'f': 'float',
+        'b': 'bool',
+        'V': 'void',
+        'O': 'object',
+        'M': 'datetime',
+        'm': 'timedelta',
+        'S': 'bytes',
+        'U': 'str',
+    }
+
+At NumPy startup, as we do today, we would generate the builtin set of
+descriptor instances, and fill in ``dtype_int_dict`` and ``dtype_scalar_type``
+so that the built-in descriptors would continue to be singletons. ``Void``,
+``Byte`` and ``Unicode`` descriptors would be constructed on demand, as is done
+today. The magic that returns a singleton or a new descriptor happens in
+``DtypeMeta.__call__``. 
+
+All descriptors would inherit from ``Dtype``::
+
+    >>> a = np.dtype(np.uint8)
+    >>> type(a).mro()
+    [<class 'UInt8Descr'>, <class 'IntDescr'>, <class 'GenericDescr'>,
+     <class 'Dtype'>, <class 'object'>]
+
+    >>> isinstance(a, np.dtype):
+    True
+    
+Note that the ``repr`` of ``a`` is compatibility with NumPy::
+
+    >>> repr(a)
+    "dtype('uint8')"
+
+Each class will have its own set of ArrFuncs (``clip``, ``fill``,
+``cast``). 
+
+Downstream users of NumPy can subclass these type classes. Creating a categorical
+dtype would look like this (without error checking for out-of-bounds values)::
+
+    class Colors(Dtype):
+        itemsize = 8
+        colors = ['red', 'green', 'blue']
+        def get_format_function(self, data, **options):
```

Comment:
Instead of formatting, I'd pick something that actually exists, perhaps the interpretation of the memory (either setting or getting).

Reply:
formatting exists. It is done as a switch on the dtype in [arrayprint](https://github.com/numpy/numpy/blob/v1.16.0rc1/numpy/core/arrayprint.py#L411)